### PR TITLE
Add common API to emit events

### PR DESCRIPTION
**Pull request by clobrano**
Add common API to emit Events to share the same format among operators.

### CODE REVIEW

File: pkg/events/events.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+)
+
+// Event message format "medik8s <operator shortname> <message>"
+const format = "medik8s %3s %s"
```
Comment (slintes) [inline pkg/events/events.go]: I think I would skip "medik8s", it's not that helpful, especially downsytream. And I would somehow separate operator and message. Maybe `"%3s: %s"`? Or `"[%3s] %s"`?

File: pkg/events/events.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+type EventRecorder struct {
+	record.EventRecorder
+	operator string
```
Comment (slintes) [inline pkg/events/events.go]: I think this needs to be public, otherwise you can´t set it when creating this struct in another package?

File: pkg/events/events.go (hunk 1 above)
Comment (clobrano) [reply pkg/events/events.go]: The idea of "medik8s" string was to help filtering our events

> Maybe "%3s: %s"? Or "[%3s] %s"?

agreed on this
EDIT, I like the latter more

File: pkg/events/events.go (hunk 2 above)
Comment (mshitrit) [reply pkg/events/events.go]: Nit: IMO might be even nicer to use a method to initialize it instead

Comment (clobrano) [reply pkg/events/events.go]: Agree on both :)

File: pkg/events/events.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+)
+
+// Event message format "medik8s <operator shortname> <message>"
+const format = "medik8s %3s %s"
+
+type EventRecorder struct {
```
Comment (mshitrit) [inline pkg/events/events.go]: IIUC this struct implements the interface of `record.EventRecorder`
it also adds 4 additional methods (NormalEvent,  WarningEvent , etc... )

- I think it should also override the original methods of `record.EventRecorder` with our formatted message.
- Nit: I'd consider returning an interface instead of a struct

File: pkg/events/events.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	// AnnotatedEventf overrides the default event recorder to add the operator shortname to the message.
+	AnnotatedEventf(object runtime.Object, annotations map[string]string, eventtype, reason, messageFmt string, args ...interface{})
+}
+
+type eventRecorder struct {
+	record.EventRecorder
+	operator string
+}
+
+var _ EventRecorder = &eventRecorder{}
```
Comment (mshitrit) [inline pkg/events/events.go]: hmm I missing something, why is this needed  ? 🤔

File: pkg/events/events.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+func (e *eventRecorder) WarningEvent(object runtime.Object, reason, message string) {
+	e.Event(object, "Warning", reason, message)
+}
+
+func (e *eventRecorder) WarningEventf(object runtime.Object, reason, messageFmt string, a ...interface{}) {
+	msg := fmt.Sprintf(messageFmt, a...)
+	e.WarningEvent(object, reason, msg)
+}
+
+func (e *eventRecorder) Event(object runtime.Object, eventtype, reason, message string) {
```
Comment (mshitrit) [inline pkg/events/events.go]: replace `eventtype` with `eventType`

File: pkg/events/events.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+)
+
+// Event message format "medik8s <operator shortname> <message>"
+const customFmt = "[%3s] %s"
```
Comment (mshitrit) [inline pkg/events/events.go]: I am missing the key word ("medik8s", "remediation" or whatever we agree on) that would help searching our events.
IIUC it should be added here instead of sent by each operator.

File: pkg/events/events.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+func (e *eventRecorder) WarningEvent(object runtime.Object, reason, message string) {
+	e.Event(object, "Warning", reason, message)
+}
+
+func (e *eventRecorder) WarningEventf(object runtime.Object, reason, messageFmt string, a ...interface{}) {
+	msg := fmt.Sprintf(messageFmt, a...)
+	e.WarningEvent(object, reason, msg)
+}
```
Comment (mshitrit) [inline pkg/events/events.go]: Why not add the formatted message with keyword and operator name to these 4 methods ? 🤔

File: pkg/events/events.go (hunk 5 above)
Comment (clobrano) [reply pkg/events/events.go]: I copy/pasted the signature from upstream, but I can change it :)

https://pkg.go.dev/k8s.io/client-go/tools/record#EventRecorder

File: pkg/events/events.go (hunk 4 above)
Comment (clobrano) [reply pkg/events/events.go]: Not needed in fact 🫣

File: pkg/events/events.go (hunk 6 above)
Comment (clobrano) [reply pkg/events/events.go]: Drop it by mistake, thanks for noticing.
At the moment we didn't completely agreed on medik8s or remediation though :shrug:

File: pkg/events/events.go (hunk 7 above)
Comment (clobrano) [reply pkg/events/events.go]: I changed the code, but for completeness :grin: I was adding the formatted message here

`e.WarningEvent` was calling the customized `Event` function which added the shared format

File: pkg/events/events_test.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+				}
+			},
+			Entry("Emit normal event",
+				nil, "Normal", "thisReason", "normal event message",
+				"Normal thisReason [remediation] normal event message"),
+			Entry("Emit normal event with args",
+				nil, "Normal", "thisReason", "normal event message with %d %s",
+				"Normal thisReason [remediation] normal event message with 2 args",
+				2, "args"),
```
Comment (mshitrit) [inline pkg/events/events_test.go]: Nit: IIUC this event message has two arguments: `2` and "args" I think using "agrs" as an argument is a bit confusing

File: pkg/events/events_test.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+			Entry("Emit warning event",
+				nil, "Warning", "thisReason", "warning event message",
+				"Warning thisReason [remediation] warning event message"),
+			Entry("Emit warning event with args",
+				nil, "Warning", "thisReason", "warning event message with %d %s",
+				"Warning thisReason [remediation] warning event message with 4 args",
+				4, "args"),
```
Comment (mshitrit) [inline pkg/events/events_test.go]: I think this test is confusing, I'm expecting four arguments for the event message but in fact there are only two: `4` and `args`

File: pkg/events/events.go [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+// Special case events
+
+// RemediationStarted will record a Normal event with reason RemediationStarted and message Remediation started.
+func RemediationStarted(recorder record.EventRecorder, object runtime.Object) {
+	NormalEvent(recorder, object, "RemediationStarted", "Remediation started")
+}
+
+// RemediationStoppedByNHC will record a Normal event with reason RemediationStopped.
+func RemediationStoppedByNHC(recorder record.EventRecorder, object runtime.Object) {
+	NormalEvent(recorder, object, "RemediationStopped", "NHC added the timed-out annotation, remediation will be stopped")
```
Comment (razo7) [inline pkg/events/events.go:45]: Is it possible to save the event's reason and message in a const which will be used here and in [`verifyEvent`](https://github.com/medik8s/machine-deletion-remediation/pull/96/files#diff-867a8834813cdf408e2387961fc75a31146379802d5abec6f465f1d8a565b903R477) function for unit tests?

Comment (clobrano) [reply pkg/events/events.go:45]: I'd rather keep it this way. I understand it creates a bit of frictions writing the tests, but I think it's a good friction. It would improve test readability (you actually read in the test source code what text is expected, without looking at the variable's value) and would spot typos easily.

Comment (razo7) [reply pkg/events/events.go:45]: I have no strong preference, fine by me.

**Comment by razo7** [comment]
/lgtm
Giving a chance to get more feedbacks, feel free to unhold
/hold

File: pkg/events/events.go [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+)
+
+// Event message format "medik8s <operator shortname> <message>"
+const customFmt = "[remediation] %s"
+
+// NormalEventf will record an event with type Normal and fixed message.
```
Comment (slintes) [inline pkg/events/events.go]: typo: NormalEvent*f*

File: pkg/events/events.go [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+)
+
+// Event message format "medik8s <operator shortname> <message>"
+const customFmt = "[remediation] %s"
```
Comment (slintes) [inline pkg/events/events.go:12]: any reason to not use this? 🤔

File: pkg/events/events.go [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+// NormalEventf will record an event with type Normal and fixed message.
+func NormalEvent(recorder record.EventRecorder, object runtime.Object, reason, message string) {
+	recorder.Event(object, "Normal", reason, fmt.Sprintf("[remediation] %s", message))
```
Comment (slintes) [inline pkg/events/events.go]: please use corev1.EventTypeNormal
(similar at other places)

File: pkg/events/events.go [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+// NormalEventf will record an event with type Normal and formatted message.
+func NormalEventf(recorder record.EventRecorder, object runtime.Object, reason, messageFmt string, a ...interface{}) {
+	message := fmt.Sprintf(messageFmt, a...)
+	recorder.Event(object, "Normal", reason, fmt.Sprintf("[remediation] %s", message))
+}
+
+// WarningEventf will record an event with type Warning and fixed message.
+func WarningEvent(recorder record.EventRecorder, object runtime.Object, reason, message string) {
```
Comment (slintes) [inline pkg/events/events.go:26]: typo: WarningEvent*f*

**Review by slintes — CHANGES_REQUESTED** [review]

File: pkg/events/events.go (hunk 12 above)
Comment (clobrano) [reply pkg/events/events.go:12]: none indeed, it was intended to be used :shrug: 
I'll fix this
